pkg/app/piped/executor/lambda/rollback.go

```diff
@@ -0,0 +1,156 @@
+	if !ok {
+		return model.StageStatus_STAGE_FAILURE
+	}
+
+	if !rollback(ctx, &e.Input, cloudProviderName, cloudProviderCfg, fm) {
+		return model.StageStatus_STAGE_FAILURE
+	}
+
+	return model.StageStatus_STAGE_SUCCESS
+}
+
+func rollback(ctx context.Context, in *executor.Input, cloudProviderName string, cloudProviderCfg *config.CloudProviderLambdaConfig, fm provider.FunctionManifest) bool {
+	in.LogPersister.Infof("Start rollback the lambda function: %s to original stage", fm.Spec.Name)
+	client, err := provider.DefaultRegistry().Client(cloudProviderName, cloudProviderCfg, in.Logger)
+	if err != nil {
+		in.LogPersister.Errorf("Unable to create Lambda client for the provider %s: %v", cloudProviderName, err)
+		return false
+	}
+
+	// Restore original traffic config from metadata store.
+	originalTrafficKeyName := fmt.Sprintf("%s-%s-original", fm.Spec.Name, in.Deployment.RunningCommitHash)
+	originalTrafficCfgData, ok := in.MetadataStore.Get(originalTrafficKeyName)
+	if !ok {
+		in.LogPersister.Errorf("Unable to prepare original traffic config to rollback Lambda function %s. It seems this is the first deployment.", fm.Spec.Name)
+		return false
+	}
+
+	originalTrafficCfg := provider.RoutingTrafficConfig{}
+	if err := json.Unmarshal([]byte(originalTrafficCfgData), &originalTrafficCfg); err != nil {
+		in.LogPersister.Errorf("Unable to prepare original traffic config to rollback Lambda function %s: %v", fm.Spec.Name, err)
+		return false
+	}
+
+	// Restore promoted traffic config from metadata store.
+	promotedTrafficKeyName := fmt.Sprintf("%s-%s-promote", fm.Spec.Name, in.Deployment.RunningCommitHash)
+	promotedTrafficCfgData, ok := in.MetadataStore.Get(promotedTrafficKeyName)
+	// If there is no previous promoted traffic config, which mean no promote run previously so no need to do anything to rollback.
+	if !ok {
+		in.LogPersister.Info("No promoted traffic config found. No need to rollback current remote traffic config.")
+		return true
+	}
+
+	promotedTrafficCfg := provider.RoutingTrafficConfig{}
+	if err := json.Unmarshal([]byte(promotedTrafficCfgData), &promotedTrafficCfg); err != nil {
+		in.LogPersister.Errorf("Unable to prepare promoted traffic config to rollback Lambda function %s: %v", fm.Spec.Name, err)
+		return false
+	}
+
+	switch len(originalTrafficCfg) {
+	// Original traffic config has both PRIMARY and SECONDARY version config.
+	case 2:
+		if err = client.UpdateTrafficConfig(ctx, fm, originalTrafficCfg); err != nil {
+			in.LogPersister.Errorf("Failed to rollback original traffic config for Lambda function %s: %v", fm.Spec.Name, err)
+			return false
+		}
+		return true
+	// Original traffic config is PRIMARY ONLY config,
+	// we need to reset any others SECONDARY created by previous (until failed) PROMOTE stages.
+	case 1:
+		// Validate stored original traffic config, since it PRIMARY ONLY, the percent must be float64(100)
+		primary, ok := originalTrafficCfg["primary"]
```

Comment:
nit: "primary" should be defined as a const.

Reply:
just to confirm, you mean the key "primary" for those maps, is that right 👀 ( so the "secondary" as well should be defined as a const too, I guess )

Reply:
Yes, the key of the map. 👍  (Because it is shared around the code.)

Reply:
🙆‍♀️

Reply:
addressed on https://github.com/pipe-cd/pipe/pull/1462/commits/a9987d0ab728505de2e3c52b9c4ea8e6560e0a9f 🙏